index: Verify the block filter hash when reading the filter from disk.

kcalvinalvin opened the pull request:
This PR picks up the abandoned #19280

BlockFilterIndex was depending on `GolombRiceDecode()` during the filter decode to sanity check that the filter wasn't corrupt. However, we can check for corruption by ensuring that the encoded blockfilter's hash matches up with the one stored in the index database.

Benchmarks that were added in #19280 showed that checking the hash is much faster.

The benchmarks were changed to nanobench and the relevant benchmarks were like below, showing a clear win for the hash check method.

```
|             ns/elem |              elem/s |    err% |        ins/elem |       bra/elem |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|---------------:|--------:|----------:|:----------
|              531.40 |        1,881,819.43 |    0.3% |        3,527.01 |         411.00 |    0.2% |      0.01 | `DecodeCheckedGCSFilter`
|          258,220.50 |            3,872.66 |    0.1% |    2,990,092.00 |     586,706.00 |    1.7% |      0.01 | `DecodeGCSFilter`
|           13,036.77 |           76,706.09 |    0.3% |       64,238.24 |         513.04 |    0.2% |      0.01 | `BlockFilterGetHash`
```

Commit:
Make sanity check in GCSFilter constructor optional

BlockFilterIndex will perform the cheaper check of verifying the filter
hash when reading the filter from disk.

Commit:
Add GCSFilterDecode and GCSBlockFilterGetHash benchmarks.

All of the benchmarks are standardized on the BASIC filter parameters
so we can compare between all the benchmarks. All the GCS
benchmarks are renamed to have "GCS" as the prefix.

Commit:
Add GCSFilterDecodeSkipCheck benchmark

This benchmark allows us to compare the differences between doing the
sanity check for corruption via GolombRiceDecode() vs checking the hash
of the encoded block filter.

jonatack commented on src/blockfilter.cpp:
```diff
@@ -59,9 +59,11 @@ GCSFilter::GCSFilter(const Params& params, std::vector<unsigned char> encoded_fi
     // Verify that the encoded filter contains exactly N elements. If it has too much or too little
     // data, a std::ios_base::failure exception will be raised.
-    BitStreamReader<SpanReader> bitreader{stream};
+    BitStreamReader<SpanReader> bitreader(stream);
```
No need to change this.

kcalvinalvin replied:
Reverted

jonatack reviewed (commented):
Concept ACK

<details><summary>A few suggestions regarding filter_checked</summary><p>

```diff
diff --git a/src/blockfilter.cpp b/src/blockfilter.cpp
index 59f34bc54e..0ab89fdbdb 100644
--- a/src/blockfilter.cpp
+++ b/src/blockfilter.cpp
@@ -47,7 +47,7 @@ GCSFilter::GCSFilter(const Params& params)
     : m_params(params), m_N(0), m_F(0), m_encoded{0}
 {}
 
-GCSFilter::GCSFilter(const Params& params, std::vector<unsigned char> encoded_filter, const bool filter_checked)
+GCSFilter::GCSFilter(const Params& params, std::vector<unsigned char> encoded_filter, bool filter_checked)
     : m_params(params), m_encoded(std::move(encoded_filter))
 {
     SpanReader stream{GCS_SER_TYPE, GCS_SER_VERSION, m_encoded};
@@ -221,7 +221,7 @@ static GCSFilter::ElementSet BasicFilterElements(const CBlock& block,
 }
 
 BlockFilter::BlockFilter(BlockFilterType filter_type, const uint256& block_hash,
-                         std::vector<unsigned char> filter, const bool filter_checked)
+                         std::vector<unsigned char> filter, bool filter_checked)
     : m_filter_type(filter_type), m_block_hash(block_hash)
 {
     GCSFilter::Params params;
diff --git a/src/blockfilter.h b/src/blockfilter.h
index af054dbf34..5de5679834 100644
--- a/src/blockfilter.h
+++ b/src/blockfilter.h
@@ -59,7 +59,7 @@ public:
     explicit GCSFilter(const Params& params = Params());
 
     /** Reconstructs an already-created filter from an encoding. */
-    GCSFilter(const Params& params, std::vector<unsigned char> encoded_filter, const bool filter_checked=false);
+    GCSFilter(const Params& params, std::vector<unsigned char> encoded_filter, bool filter_checked = false);
 
     /** Builds a new filter from the params and set of elements. */
     GCSFilter(const Params& params, const ElementSet& elements);
@@ -122,7 +122,7 @@ public:
 
     //! Reconstruct a BlockFilter from parts.
     BlockFilter(BlockFilterType filter_type, const uint256& block_hash,
-                std::vector<unsigned char> filter, const bool filter_checked=false);
+                std::vector<unsigned char> filter, bool filter_checked = false);
 
     //! Construct a new BlockFilter of the specified type from a block.
     BlockFilter(BlockFilterType filter_type, const CBlock& block, const CBlockUndo& block_undo);
diff --git a/src/index/blockfilterindex.cpp b/src/index/blockfilterindex.cpp
index bad766ebe1..0a4728f571 100644
--- a/src/index/blockfilterindex.cpp
+++ b/src/index/blockfilterindex.cpp
@@ -159,7 +159,7 @@ bool BlockFilterIndex::ReadFilterFromDisk(const FlatFilePos& pos, BlockFilter& f
         uint256 result;
         CHash256().Write(encoded_filter).Finalize(result);
         if (result != hash) return error("Checksum mismatch in filter decode.");
-        filter = BlockFilter(GetFilterType(), block_hash, std::move(encoded_filter), true);
+        filter = BlockFilter(GetFilterType(), block_hash, std::move(encoded_filter), /*filter_checked=*/true);
     }
```
</p></details>

kcalvinalvin commented:
> A few suggestions regarding filter_checked

I've applied all the suggested changes :)

jonatack commented on src/bench/gcs_filter.cpp:
```diff
@@ -32,12 +32,65 @@ static void MatchGCSFilter(benchmark::Bench& bench)
+    });
+}
+
+static void BlockFilterGetHash(benchmark::Bench& bench)
```
It would be handy to name all of these benchmarks with a common prefix to group them together in the output and that is easy to filter for, i.e. `./src/bench/bench_bitcoin -filter=EvictionProtection*.*`

kcalvinalvin replied:
All the GCSFIlter benchmarks are now changed to have the prefix `GCSFilter`.

jonatack commented on src/bench/gcs_filter.cpp:
```diff
@@ -32,12 +32,65 @@ static void MatchGCSFilter(benchmark::Bench& bench)
+        GCSFilter::Element element(32);
+        element[0] = static_cast<unsigned char>(i);
+        element[1] = static_cast<unsigned char>(i >> 8);
+        elements.insert(std::move(element));
```
It looks like there is a fair amount of duplicate boilerplate in each of these benchmarks that you could extract out to a common setup helper in a separate commit, if you like. See  `EvictionProtectionCommon()` in `src/bench/peer_eviction.cpp` for an example.

kcalvinalvin replied:
Since most of the duplicate code is just generating  the elements, how's separating that out to a function like so:

```
static const GCSFilter::ElementSet GenerateGCSTestElements(){
    GCSFilter::ElementSet elements;
    for (int i = 0; i < 10000; ++i) {
        GCSFilter::Element element(32);
        element[0] = static_cast<unsigned char>(i);
        element[1] = static_cast<unsigned char>(i >> 8);
        elements.insert(std::move(element));
    }

    return elements;
}
```

And having each of the benchmarks call that function?

kcalvinalvin replied:
I've applied the above change I suggested above.

jonatack commented on src/blockfilter.h:
```diff
@@ -59,7 +59,7 @@ class GCSFilter
 
     /** Reconstructs an already-created filter from an encoding. */
-    GCSFilter(const Params& params, std::vector<unsigned char> encoded_filter);
+    GCSFilter(const Params& params, std::vector<unsigned char> encoded_filter, bool filter_checked=false);
```
Clang format for each of these defaults
```suggestion
    GCSFilter(const Params& params, std::vector<unsigned char> encoded_filter, bool filter_checked = false);
```

kcalvinalvin replied:
No longer needed as the default value was removed

jonatack reviewed (commented):
Approach ACK, a few iterative further comments on second look.

jonatack commented on src/bench/gcs_filter.cpp:
```diff
@@ -32,12 +32,65 @@ static void MatchGCSFilter(benchmark::Bench& bench)
+    auto encoded = filter.GetEncoded();
+
+    bench.unit("elem").run([&] {
+        GCSFilter filter({0, 0, BASIC_FILTER_P, BASIC_FILTER_M}, encoded, true);
```
```suggestion
        GCSFilter filter{{0, 0, BASIC_FILTER_P, BASIC_FILTER_M}, encoded, /*filter_checked=*/true};
```

kcalvinalvin replied:
Addressed

mzumsande reviewed (commented):
Concept ACK

As far as I can see, the sanity check that this PR makes optional is now only invoked from `BlockFilter::Unserialize()`, which is used only in test code because as a full node, bitcoin core doesn't have the use case of accepting unchecked serialized blockfilters over p2p, only send self-generated ones to peers. 
So I wonder whether it might make sense to just remove the old sanity check instead of making it optional?

jonatack commented:
@kcalvinalvin maybe decided case-by-case; I don't recall a recent specific example of leave-it-in-for-reuse-as-a-library being invoked as a rationale for not moving code used only for tests out to the test code.

Verified that removing `BlockFilter::Unserialize()` compiles with the following test-only deletions:

<details><summary>code that depends on BlockFilter::Unserialize()</summary><p>

```diff
diff --git a/src/blockfilter.h b/src/blockfilter.h
index 21d6a295a2..764395c370 100644
--- a/src/blockfilter.h
+++ b/src/blockfilter.h
@@ -148,24 +148,6 @@ public:
           << m_block_hash
           << m_filter.GetEncoded();
     }
-
-    template <typename Stream>
-    void Unserialize(Stream& s) {
-        std::vector<unsigned char> encoded_filter;
-        uint8_t filter_type;
-
-        s >> filter_type
-          >> m_block_hash
-          >> encoded_filter;
-
-        m_filter_type = static_cast<BlockFilterType>(filter_type);
-
-        GCSFilter::Params params;
-        if (!BuildParams(params)) {
-            throw std::ios_base::failure("unknown filter_type");
-        }
-        m_filter = GCSFilter(params, std::move(encoded_filter));
-    }
 };
 
 #endif // BITCOIN_BLOCKFILTER_H
diff --git a/src/test/blockfilter_tests.cpp b/src/test/blockfilter_tests.cpp
index 8eb4dbc592..b99d9a32c3 100644
--- a/src/test/blockfilter_tests.cpp
+++ b/src/test/blockfilter_tests.cpp
@@ -106,23 +106,6 @@ BOOST_AUTO_TEST_CASE(blockfilter_basic_test)
     for (const CScript& script : excluded_scripts) {
         BOOST_CHECK(!filter.Match(GCSFilter::Element(script.begin(), script.end())));
     }
-
-    // Test serialization/unserialization.
-    BlockFilter block_filter2;
-
-    CDataStream stream(SER_NETWORK, PROTOCOL_VERSION);
-    stream << block_filter;
-    stream >> block_filter2;
-
-    BOOST_CHECK_EQUAL(block_filter.GetFilterType(), block_filter2.GetFilterType());
-    BOOST_CHECK_EQUAL(block_filter.GetBlockHash(), block_filter2.GetBlockHash());
-    BOOST_CHECK(block_filter.GetEncodedFilter() == block_filter2.GetEncodedFilter());
-
-    BlockFilter default_ctor_block_filter_1;
-    BlockFilter default_ctor_block_filter_2;
-    BOOST_CHECK_EQUAL(default_ctor_block_filter_1.GetFilterType(), default_ctor_block_filter_2.GetFilterType());
-    BOOST_CHECK_EQUAL(default_ctor_block_filter_1.GetBlockHash(), default_ctor_block_filter_2.GetBlockHash());
-    BOOST_CHECK(default_ctor_block_filter_1.GetEncodedFilter() == default_ctor_block_filter_2.GetEncodedFilter());
 }
 
 BOOST_AUTO_TEST_CASE(blockfilters_json_test)
diff --git a/src/test/fuzz/deserialize.cpp b/src/test/fuzz/deserialize.cpp
index ed6f172a2a..3793921e86 100644
--- a/src/test/fuzz/deserialize.cpp
+++ b/src/test/fuzz/deserialize.cpp
@@ -95,12 +95,6 @@ void DeserializeFromFuzzingInput(FuzzBufferType buffer, T& obj, const std::optio
             throw invalid_fuzzing_input_exception();
         }
     }
-    try {
-        ds >> obj;
-    } catch (const std::ios_base::failure&) {
-        throw invalid_fuzzing_input_exception();
-    }
-    assert(buffer.empty() || !Serialize(obj).empty());
 }
 
 template <typename T>
diff --git a/src/test/fuzz/util.h b/src/test/fuzz/util.h
index 6c91844633..e8e8e57394 100644
--- a/src/test/fuzz/util.h
+++ b/src/test/fuzz/util.h
@@ -145,11 +145,6 @@ template <typename T>
     const std::vector<uint8_t> buffer = ConsumeRandomLengthByteVector(fuzzed_data_provider, max_length);
     CDataStream ds{buffer, SER_NETWORK, INIT_PROTO_VERSION};
     T obj;
-    try {
-        ds >> obj;
-    } catch (const std::ios_base::failure&) {
-        return std::nullopt;
-    }
     return obj;
 }
```
</p></details>

kcalvinalvin commented on src/index/blockfilterindex.h:
```diff
@@ -31,7 +31,7 @@ class BlockFilterIndex final : public BaseIndex
     std::unique_ptr<FlatFileSeq> m_filter_fileseq;
 
-    bool ReadFilterFromDisk(const FlatFilePos& pos, BlockFilter& filter) const;
+    bool ReadFilterFromDisk(const FlatFilePos& pos, BlockFilter& filter, const uint256& hash) const;
```
Addressed

kcalvinalvin commented:
Addressed all of the changes requested except for removing `BlockFilter::Unserialize()` and getting rid of the option to sanity check. I'll do those in a follow up.

jonatack commented:
ACK 3e6d98e2dc98a4305a5bbdf4b6bea10bcf4aca71

modulo:
- ordering the `gcs_filter` benchmark methods in the same order as the benchmarks in that file
- maybe naming the `BlockFilterGetHash` bench as `GCSBlockFilterGetHash` so that `./src/bench/bench_bitcoin -filter=GCS*.*` can invoke it with the others in that file

Here is a patch on top of your last push, if helpful, that removes `BlockFilter::Unserialize`: https://github.com/jonatack/bitcoin/commits/remove-BlockFilter-Unserialize

jonatack commented:
ACK 0b976a6dc0003898abaef14254f572b58f83926a

Sanity check:

Running `./src/bench/bench_bitcoin -filter=GCS*.*` on this branch (debug build and not tuned for benchmarking)...
```
|             ns/elem |              elem/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          179,394.67 |            5,574.30 |    0.5% |      0.01 | `GCSBlockFilterGetHash`
|            3,376.66 |          296,150.75 |    2.1% |      0.38 | `GCSFilterConstruct`
|        5,163,985.00 |              193.65 |    2.5% |      0.06 | `GCSFilterDecode`
|            1,926.40 |          519,102.50 |    1.9% |      0.01 | `GCSFilterDecodeChecked`
|          577,446.00 |            1,731.76 |    1.0% |      0.01 | `GCSFilterMatch`
```
...and with the optimisation commented out...
```diff
--- a/src/blockfilter.cpp
+++ b/src/blockfilter.cpp
@@ -59,7 +59,7 @@ GCSFilter::GCSFilter(const Params& params, std::vector<unsigned char> encoded_fi
     }
     m_F = static_cast<uint64_t>(m_N) * static_cast<uint64_t>(m_params.m_M);
 
-    if (filter_checked) return;
+    // if (filter_checked) return;
```
...GCSFilterDecodeChecked is far slower and equivalent to GCSFilterDecode as expected
```
|             ns/elem |              elem/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          155,888.33 |            6,414.85 |    0.6% |      0.01 | `GCSBlockFilterGetHash`
|            3,859.72 |          259,086.04 |    2.3% |      0.43 | `GCSFilterConstruct`
|        4,933,143.00 |              202.71 |    0.7% |      0.06 | `GCSFilterDecode`
|        4,955,108.00 |              201.81 |    0.7% |      0.05 | `GCSFilterDecodeChecked`
|          562,882.00 |            1,776.57 |    0.5% |      0.01 | `GCSFilterMatch`
```

kcalvinalvin commented:
> ACK [0b976a6](https://github.com/bitcoin/bitcoin/commit/0b976a6dc0003898abaef14254f572b58f83926a)

Thank you for the speedy review!

I pushed d492499e950bdcbe2a878c9488b3607acdec3dc7 on top of 0b976a6dc0003898abaef14254f572b58f83926a which does mostly what you did in https://github.com/jonatack/bitcoin/commits/remove-BlockFilter-Unserialize but it keeps the serialize/unserialize test in `src/test/blockfilter_tests.cpp` with the function

```
template <typename Stream>
static BlockFilter UnserializeBlockFilter(Stream& s) {
    std::vector<unsigned char> encoded_filter;
    uint8_t filter_type_uint8;
    uint256 block_hash;

    s >> filter_type_uint8
      >> block_hash
      >> encoded_filter;

    BlockFilterType filter_type = static_cast<BlockFilterType>(filter_type_uint8);
    BlockFilter block_filter(filter_type, block_hash, std::move(encoded_filter), /*filter_checked=*/false);
    return block_filter;
}
```

I think this is a good tradeoff between keeping the serialization in the unit tests but still keeping the maintenance overhead low.

kcalvinalvin commented:
> Looks like you are removing the fuzz test as well?

@MarcoFalke Ah my first instinct was to get rid of the entire test as `ConsumeDeserializable` would no longer be usable. I'll push a change with the fuzz tests kept in.

jonatack commented:
@kcalvinalvin if you remove BlockFilter::Unserialize() in this pull, it will be a smaller change as the first commit rather than last one, WDYT?

kcalvinalvin commented:
> @kcalvinalvin if you move BlockFilter::Unserialize() to the test code in this pull, it will be a smaller change as the first commit rather than last one, WDYT? Edit: maybe a first commit that adds the check in ReadFilterFromDisk() and rename the commit, then the one that moves Unserialize() to the test code.

I'm ok with that if it'd lower the cost of reviewing. However, I'm thinking if the cost of reviewing would be even lower if I have commit 447174d208bc3fdd32874656344a736edf949e63 be in a separate PR. I'm sorta thinking maybe getting rid of `Unserialize()` and replacing `GolombRiceDecode()` check are two separate things. Let me know what you'd think would lower the cost of reviewing :)

jonatack commented:
@kcalvinalvin SGTM.

kcalvinalvin commented:
Reverted remove the last commit. Latest commit is now back to 0b976a6dc0003898abaef14254f572b58f83926a. Will make a follow up PR after this one.

ryanofsky commented on src/blockfilter.cpp:
```diff
@@ -47,7 +47,7 @@ GCSFilter::GCSFilter(const Params& params)
 {}
 
-GCSFilter::GCSFilter(const Params& params, std::vector<unsigned char> encoded_filter)
+GCSFilter::GCSFilter(const Params& params, std::vector<unsigned char> encoded_filter, bool filter_checked)
```
In commit "Make sanity check in GCSFilter constructor optional" (c35df324b19b94aaecfbcd26111b7d00c5bed4e0)

Requesting change: Can we rename `filter_checked` to `skip_checks` or `skip_decode_check`. I spent an embarrassing amount of time being confused by the changes in this PR because I interpreted `filter_checked` to mean "filter should be checked" not "filter should not be checked" which is what it seems to actually mean.

kcalvinalvin replied:
Changed `filter_checked` to `skip_decode_check`

ryanofsky reviewed (approved):
Code review ACK 0b976a6dc0003898abaef14254f572b58f83926a

ryanofsky reviewed (approved):
Code review ACK b97685965c6646b3a8521eb49d1f79b03a1b473f. Just renamed `filter_checked` variable since last review (thanks!)

jonatack reviewed (approved):
Diff-review-only ACK b97685965c6646b3a8521eb49d1f79b03a1b473f per `git diff 0b976a6 b976859`, only change since my previous full review at https://github.com/bitcoin/bitcoin/pull/24832#issuecomment-1102816767 is renaming the bool param added to ` GCSFilter()` from `filter_checked` to `skip_decode_check` -- this param is expected to go away in the planned follow-up commit 04008e2df6e2bac0cc92a2be05fec257a0fe7d7f to remove `BlockFilter::Unserialize()`

mzumsande commented on src/bench/gcs_filter.cpp:
```diff
@@ -15,29 +15,56 @@ static void ConstructGCSFilter(benchmark::Bench& bench)
+    return elements;
+}
+
+static void GCSBlockFilterGetHash(benchmark::Bench& bench)
```
Could name it GCSFilterGetHash instead of GCSBlockFilterGetHash for consistency with the other benchmarks.

kcalvinalvin commented on src/bench/gcs_filter.cpp:
```diff
@@ -15,29 +15,56 @@ static void ConstructGCSFilter(benchmark::Bench& bench)
+    GCSFilter filter({0, 0, BASIC_FILTER_P, BASIC_FILTER_M}, elements);
+    BlockFilter block_filter(BlockFilterType::BASIC, {}, filter.GetEncoded(), /*skip_decode_check=*/false);
+
+    bench.unit("elem").run([&] {
```
Will change all the units to `op` except for `GCSFilterConstruct` as that benchmark still is done with the 10k batch.

mzumsande replied:
no need to type "op" (it's the default), could just remove the `.unit(...)` as in most other benchmarks.

kcalvinalvin replied:
Also fixed

mzumsande commented:
Code review ACK b32e96578182b9af3283c12829d080fc4bba2be4

jonatack commented:
<details><summary>I'd like to propose this diff. Rationale: compare all operations on the same basis (per operation).</summary><p>

```diff
--- a/src/bench/gcs_filter.cpp
+++ b/src/bench/gcs_filter.cpp
@@ -5,10 +5,12 @@
 #include <bench/bench.h>
 #include <blockfilter.h>
 
+static constexpr int NUM_ELEMENTS = 10'000;
+
 static const GCSFilter::ElementSet GenerateGCSTestElements()
 {
     GCSFilter::ElementSet elements;
-    for (int i = 0; i < 10000; ++i) {
+    for (int i = 0; i < NUM_ELEMENTS; ++i) {
         GCSFilter::Element element(32);
         element[0] = static_cast<unsigned char>(i);
         element[1] = static_cast<unsigned char>(i >> 8);
@@ -18,15 +20,27 @@ static const GCSFilter::ElementSet GenerateGCSTestElements()
     return elements;
 }
 
+static void GCSBlockFilterGetHash(benchmark::Bench& bench)
+{
+    auto elements = GenerateGCSTestElements();
+
+    GCSFilter filter({0, 0, BASIC_FILTER_P, BASIC_FILTER_M}, elements);
+    BlockFilter block_filter(BlockFilterType::BASIC, {}, filter.GetEncoded(), /*skip_decode_check=*/false);
+
+    bench.batch(NUM_ELEMENTS).run([&] {
+        block_filter.GetHash();
+    });
+}
+
 static void GCSFilterConstruct(benchmark::Bench& bench)
 {
     auto elements = GenerateGCSTestElements();
 
     uint64_t siphash_k0 = 0;
-    bench.batch(elements.size()).unit("elem").run([&] {
+    bench.batch(NUM_ELEMENTS).run([&] {
         GCSFilter filter({siphash_k0, 0, BASIC_FILTER_P, BASIC_FILTER_M}, elements);
 
-        siphash_k0++;
+        ++siphash_k0;
     });
 }
 
@@ -37,7 +51,7 @@ static void GCSFilterDecode(benchmark::Bench& bench)
     GCSFilter filter({0, 0, BASIC_FILTER_P, BASIC_FILTER_M}, elements);
     auto encoded = filter.GetEncoded();
 
-    bench.run([&] {
+    bench.batch(NUM_ELEMENTS).run([&] {
         GCSFilter filter({0, 0, BASIC_FILTER_P, BASIC_FILTER_M}, encoded, /*skip_decode_check=*/false);
     });
 }
@@ -49,35 +63,24 @@ static void GCSFilterDecodeChecked(benchmark::Bench& bench)
     GCSFilter filter({0, 0, BASIC_FILTER_P, BASIC_FILTER_M}, elements);
     auto encoded = filter.GetEncoded();
 
-    bench.run([&] {
+    bench.batch(NUM_ELEMENTS).run([&] {
         GCSFilter filter({0, 0, BASIC_FILTER_P, BASIC_FILTER_M}, encoded, /*skip_decode_check=*/true);
     });
 }
 
-static void GCSFilterGetHash(benchmark::Bench& bench)
-{
-    auto elements = GenerateGCSTestElements();
-
-    GCSFilter filter({0, 0, BASIC_FILTER_P, BASIC_FILTER_M}, elements);
-    BlockFilter block_filter(BlockFilterType::BASIC, {}, filter.GetEncoded(), /*skip_decode_check=*/false);
-
-    bench.run([&] {
-        block_filter.GetHash();
-    });
-}
-
 static void GCSFilterMatch(benchmark::Bench& bench)
 {
     auto elements = GenerateGCSTestElements();
 
     GCSFilter filter({0, 0, BASIC_FILTER_P, BASIC_FILTER_M}, elements);
 
-    bench.run([&] {
+    bench.batch(NUM_ELEMENTS).run([&] {
         filter.Match(GCSFilter::Element());
     });
 }
+
+BENCHMARK(GCSBlockFilterGetHash);
 BENCHMARK(GCSFilterConstruct);
 BENCHMARK(GCSFilterDecode);
 BENCHMARK(GCSFilterDecodeChecked);
-BENCHMARK(GCSFilterGetHash);
 BENCHMARK(GCSFilterMatch);
```
</p></details>

Output of last push

```
|             ns/elem |              elem/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|            2,257.12 |          443,042.46 |    3.1% |      0.25 | `GCSFilterConstruct`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|        2,800,765.00 |              357.05 |    3.1% |      0.03 | `GCSFilterDecode`
|            1,218.41 |          820,739.84 |   11.4% |      0.01 | `GCSFilterDecodeChecked`
|           93,894.33 |           10,650.27 |    4.1% |      0.01 | `GCSFilterGetHash`
|          315,092.00 |            3,173.68 |    2.7% |      0.01 | `GCSFilterMatch`
```

Output with proposed diff

```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                9.02 |      110,833,467.68 |    0.0% |      0.01 | `GCSBlockFilterGetHash`
|            2,183.44 |          457,992.58 |    1.3% |      0.24 | `GCSFilterConstruct`
|              264.56 |        3,779,922.34 |    1.7% |      0.03 | `GCSFilterDecode`
|                0.11 |    9,386,117,464.60 |    1.6% |      0.01 | `GCSFilterDecodeChecked`
|               30.60 |       32,684,901.21 |    1.6% |      0.01 | `GCSFilterMatch`
```

Other notes:

- Could be wrong but it seems to me the `GCSBlockFilterGetHash` name is preferable, as that describes the particularity of that bench.
- I'm indifferent to "op" or "elem"; "op" for all the benchmarks here is fine.

mzumsande commented:
> Could be wrong but it seems to me the GCSBlockFilterGetHash name is preferable, as that describes the particularity of that bench.

No, you are right, I missed before that this benchmark is on the level of `BlockFilter`, whereas the others are on the level of `GCSFilter`, so I agree.

> Rationale: compare all operations on the same basis (per operation).

> I'm indifferent to "op" or "elem"; "op" for all the benchmarks here is fine.

My understanding of the bench framework (correct me if I'm wrong!) is that  `ns/op=10000` should be read such that one operation (whatever is inside of `run()`) takes 10000 ns on average.
If we then choose `.batch(NUM_ELEMENTS)`, we'd expect this number to be divided by NUM_ELEMENTS and thus go down to `ns/elem=1` on average. So this means that it takes 1ns to process 1 filter element (so using the default ns/op would be kind of misleading in combination with `.batch()`). 
For this to make sense, there should be a linear relation between the custom unit ("elements") and the time, so if we used `.batch(NUM_ELEMENTS)` and set NUM_ELEMENTS=100,000 instead of 10,000  the benchmark results shouldn't change - which seems not really the case for some of the benchmarks here, at least locally for me.
So maybe using the default `ns/op` and no `.batch()` for each benchmark would be the best here? Does that make sense?

jonatack commented:
I need to verify after tuning for benchmarking. My hypothesis (to be confirmed or not) is that `batch` appears to behave like a linear scale factor. I agree that it seems good to use `batch` for all of them (result per op) or for none (result per batched op) to make comparing intra-bench results more intuitive.

jonatack commented:
(Also to reverify: lower err% with `batch`?)

kcalvinalvin commented:
Changing `NUM_ELEMENTS` to 100,000 from 10,000 results in different ns/elem with my untuned system. I'll try running the benchmark with a tuned system.

kcalvinalvin commented:
# My system

I got around to doing the benchmarks with a (sorta) tuned system. I wasn't able to achieve a completely stable benchmark system but the error rates are much lower now than before.

output of `sudo pyperf system`:
```
[I] calvin@bitcoin ~/b/c/l/bitcoin (2020-06-14-blockfilterindex-checksums) [1]> sudo pyperf system
Show the system configuration

System state
============

CPU: use 6 logical CPUs: 0-5
Perf event: Maximum sample rate: 1 per second
ASLR: Full randomization
Linux scheduler: Isolated CPUs (6/12): 0-5
Linux scheduler: RCU disabled on CPUs (6/12): 0-5
CPU Frequency: 0-5=min=max=3600 MHz; 6-11=min=2200 MHz, max=3600 MHz
IRQ affinity: irqbalance service: inactive
IRQ affinity: Default IRQ affinity: CPU 6-11
IRQ affinity: IRQ affinity: IRQ 0,2=CPU 0-11; IRQ 1,3-15,25,40-41,73,75,77-78,80-89,91,93=CPU 6-11; IRQ 42,57=CPU 0; IRQ 43,58=CPU 1; IRQ 44,59=CPU 2; IRQ 45,60=CPU 3; IRQ 46,61=CPU 4; IRQ 47,62=CPU 5; IRQ 48,63=CPU 6; IRQ 49,64=CPU 7; IRQ 50,65=CPU 8; IRQ 51,66=CPU 9; IRQ 52,67=CPU 10; IRQ 53,68=CPU 11; IRQ 54-56,69-72=CPU None

Warnings
========

Turbo Boost (MSR): Failed to read MSR 0x1a0 from /dev/cpu/0/msr: [Errno 5] Input/output error

OK! System ready for benchmarking
```

For some reason pyperf doesn't seem to recognize amd cpus with boost turned off. Boost is manually turned off by writing 0 to /sys/devices/system/cpu/cpufreq/boost.

`cat /sys/devices/system/cpu/cpufreq/boost` output:
```
[N] calvin@bitcoin ~/b/c/l/bitcoin (2020-06-14-blockfilterindex-checksums)> cat /sys/devices/system/cpu/cpufreq/boost
0
```

I get this output when running the benchmarks:
```
Warning, results might be unstable:
* CPU frequency scaling enabled: CPU 6 between 2,200.0 and 3,600.0 MHz
* Turbo is enabled, CPU frequency will fluctuate

Recommendations
* Use 'pyperf system tune' before benchmarking. See https://github.com/psf/pyperf
```
But I believe pyperf only uses the cpus that are isolated from the kernel scheduler so I believe the warning will not affect the benchmarks. I double checked and also isolated CPU 6 and pyperf repeated the same warning with CPU 7 so I think it's ok. Pyperf also doesn't support turning off boost for [amd cpus](https://github.com/psf/pyperf/issues/120) so I believe that warning is an error by pyperf.

# Benchmarks

I tested to see if the results would be linear across many different `NUM_ELEMENT`s. Since the original benchmarks were done with 10,000, I took powers of 10 for the `NUM_ELEMENT`s for my test, ranging from 1-100,000,000.

The benchmarks were done with batching and without batching. The batched test resulted in a graph like so:

![results_from_benchmarks_ns-elem](https://user-images.githubusercontent.com/37185887/169595920-c68b1373-0f7d-4799-8c5f-c421fc1467ff.png)

The tests done without batching resulted in a graph like so:

![results_from_benchmarks_ns-op](https://user-images.githubusercontent.com/37185887/169595980-a78ae29a-93e0-4909-933e-0d43f152b8f4.png)

From these benchmarks, you can see that there is a linear relation between the custom unit ("elements") and the time once you have 100,000 elements. Anything below 100,000 doesn't result in a linear relation.

`ns/elem` is easier to read than `ns/op` since 9.7million is kinda hard to read. However, you can also take a look at `op/s` which is just as easy to read as `ns/elem`. Here's a graph without batch with `op/s` instead of `ns/op`.

![results_from_benchmarks_op-s](https://user-images.githubusercontent.com/37185887/169598934-e31d957d-9901-48e5-9867-5b49674843ec.png)

# Error rates with and without batch

I also graphed how the error rates compare between calling batch and not calling batch.

![error_rates_for_benchmarks](https://user-images.githubusercontent.com/37185887/169596271-df172ec1-83c9-42a1-9b66-0da8ee6157ec.png)

Average error rate in % with batch:       `0.38888888888888884`
Average error rate in % without batch: `0.4`

Calling batch doesn't seem to effect the error rates in my testing.

# My takeaway

- Lower `NUM_ELEMENT` count seem to result in ns/elem that's same for different `NUM_ELEMENT` values. Needs a minimum of 100,000 elements.
- It'd be better to unify everything under `ns/op` as batch doesn't provide any less error rates.

[txt file for the raw benchmark results](https://github.com/bitcoin/bitcoin/files/8743561/raw-benchmark-num.txt)

kcalvinalvin committed:
Update GCSFilter benchmarks

Element count used in the GCSFilter benchmarks are increased to 100,000
from 10,000. Testing the benchmarks with different element counts showed
that a filter with 100,000 elements resulted in the same ns/op. This
this a desirable thing to have as it allows us to reason about how long
a single filter element takes to process, letting us easily calculate
how long a filter with N elements (where N > 100,000) would take to
process.

GCSFilterConstruct benchmark is now called without batch. This makes
intra-bench results more intuitive as all benchmarks are in ns/op
instead of a custom unit. There are no downsides to this change as
testing showed that there is no observable difference in error rates
in the benchmarks when calling without batch.

kcalvinalvin commented:
3 things changed in the latest push:

1. Renamed the `GCSFilterGetHash` benchmark back to `GCSBlockFilterGetHash`.
2. Renamed `GCSFilterDecodeChecked` to `GCSFilterDecodeSkipCheck` as `filter_checked` argument was renamed to `skip_decode_check`. Changing the name of the benchmark as well seemed fitting.
3. Added a commit where `GenerateGCSTestElements` generates 100,000 elements instead of 10,000. The reasoning was based off of my findings in testing the benchmarks. I've included an explanation in the code as a comment and in the commit message.
4. In the same added commit, I've changed the `GCSFilterConstruct` to be called without batch and the custom unit. This was done as it makes comparing intra-bench results more intuitive. Calling with batch didn't result in lower error rates either so there were no tradeoffs made.

mzumsande reviewed (commented):
Code Review ACK e734228d8585c0870c71ce8ba8c037f8cf8b249a

I'm not sure if it really matters a lot how big we choose the number of elements used in the test, I just think that if we pick custom measure other than "ops", the benchmark should be linear in that - so I like that everything uses `ns/op` now.
In any case, the important feature of this PR is the improvement of the filter verification.

theStack commented:
Concept ACK

theStack reviewed (approved):
Code-review ACK e734228d8585c0870c71ce8ba8c037f8cf8b249a

FWIW, this are the benchmark results on my machine:
```
$ ./src/bench/bench_bitcoin -filter=GCS.*
```

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          719,504.00 |            1,389.85 |    2.4% |      0.01 | `GCSBlockFilterGetHash`
|       28,431,064.00 |               35.17 |    4.0% |      0.31 | `GCSFilterConstruct`
|        3,641,177.00 |              274.64 |    3.4% |      0.04 | `GCSFilterDecode`
|           13,994.04 |           71,458.99 |    1.5% |      0.01 | `GCSFilterDecodeSkipCheck`
|          424,500.50 |            2,355.71 |    2.2% |      0.01 | `GCSFilterMatch`

stickies-v reviewed (approved):
ACK e734228d8585c0870c71ce8ba8c037f8cf8b249a

Non-controversial and significant performance improvement when loading block filters from disk, which can be a frequent process depending on peers or RPC users. I've left some style/readability suggestions but none of them are blocking for me.

If `BlockFilter::Unserialize()` is removed in a follow up, further code cleanup is possible by removing the `skip_decode_check` parameter and the check branch.

`$ ./src/bench/bench_bitcoin -filter=GCS.*`
yields
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          766,333.00 |            1,304.92 |    0.8% |      0.01 | `GCSBlockFilterGetHash`
|       60,891,167.00 |               16.42 |    0.3% |      0.67 | `GCSFilterConstruct`
|       15,802,125.00 |               63.28 |    0.3% |      0.17 | `GCSFilterDecode`
|        1,697,875.00 |              588.97 |    0.2% |      0.02 | `GCSFilterDecodeSkipCheck`
|        1,687,000.00 |              592.77 |    0.1% |      0.02 | `GCSFilterMatch`
```

ryanofsky reviewed (approved):
Code review ACK e734228d8585c0870c71ce8ba8c037f8cf8b249a, with caveat that I mostly paid attention to the main code, not the changes to the benchmark. Only changes since last review were changes to the benchmark code.

fanquake commented:
@stickies-v maybe you'd like to follow up here?

kcalvinalvin commented:
I have working code here: https://github.com/kcalvinalvin/bitcoin/commits/2020-06-14-blockfilterindex-checksums-2. I'll make a follow up PR

jonatack commented:
Post-merge re-ACK e734228d8585c0870c71ce8ba8c037f8cf8b249a, thanks for updating.
